### PR TITLE
Re-introducing Ml ai feature to 13.0

### PR DESCRIPTION
**Pull request by alex-rakowski**

### CODE REVIEW

**Comment by alex-rakowski** [comment]
This is in a working state, and can be tired against below:
To do:
- [ ] handle pass-through of ml variables  
- [ ] check output of peak detection are correct data class 
- [ ] remove any duplication


```
import py4DSTEM
import numpy as np
file_path_data = 'small_AuAgPd_wire_dataset_04_3.h5'
file_path_probe = 'downsampled_AuAgPd_wire_probe_1.h5'
file_path_analysis = '/AuAgPd_wire_analysis_'


# load data
dataset = py4DSTEM.io.read(file_path_data, data_id='datacube_0')
probe = py4DSTEM.io.read(file_path_probe, data_id='datacube_0', )
# pad qspace to (256,256) and crop to real space to speed up 



dataset.data = np.pad(dataset.data,((0,0),(0,0),(42,42),(42,42)),mode='constant',constant_values=0)[:10,:10]
probe.data = np.pad(probe.data,((0,0),(0,0),(42,42),(42,42)),mode='constant',constant_values=0)

print(dataset.data.shape)
print(probe.data.shape)


peaks = dataset.find_Bragg_disks(
    template=probe.data.mean((0,1)),
    ML=True,
    CUDA=False,
    ml_num_attempts=1
)

# check output types and if added to datacube correctly
print(peaks)
print(dataset.tree)

```

File: py4DSTEM/process/diskdetection/diskdetection.py [hunk 1]
```diff
@@ -206,8 +214,13 @@ def find_Bragg_disks(
...
                 er = f"unrecognized distributed mode {distributed_mode}"
                 raise Exception(er)
+    # overwrite if ML selected
+    if ML:
+        mode = 'dc_ml_ai'
+        # need to check that the probe is in real space
+        assert np.isrealobj(template), "Template dtype not real"
 
-
+    print(mode)
```
Comment (alex-rakowski) [inline py4DSTEM/process/diskdetection/diskdetection.py:223]: remove this

File: py4DSTEM/process/diskdetection/diskdetection.py [hunk 2]
```diff
@@ -206,8 +214,13 @@ def find_Bragg_disks(
             else:
                 er = f"unrecognized distributed mode {distributed_mode}"
                 raise Exception(er)
+    # overwrite if ML selected
+    if ML:
+        mode = 'dc_ml_ai'
+        # need to check that the probe is in real space
+        assert np.isrealobj(template), "Template dtype not real"
```
Comment (alex-rakowski) [inline py4DSTEM/process/diskdetection/diskdetection.py:221]: move this above and set next block to `elif`

File: py4DSTEM/process/diskdetection/diskdetection.py [hunk 3]
```diff
@@ -206,8 +214,13 @@ def find_Bragg_disks(
             else:
                 er = f"unrecognized distributed mode {distributed_mode}"
                 raise Exception(er)
+    # overwrite if ML selected
+    if ML:
+        mode = 'dc_ml_ai'
+        # need to check that the probe is in real space
```
Comment (alex-rakowski) [inline py4DSTEM/process/diskdetection/diskdetection.py:220]: change to `dc_ml` and update function dict

File: py4DSTEM/process/diskdetection/diskdetection.py (hunk 2 above)
Comment (alex-rakowski) [inline py4DSTEM/process/diskdetection/diskdetection.py:221]: remove as doesn't do what I wanted

File: py4DSTEM/process/diskdetection/diskdetection.py [hunk 4]
```diff
@@ -222,6 +235,12 @@ def find_Bragg_disks(
         kws['connect'] = connect
         kws['data_file'] = data_file
         kws['cluster_path'] = cluster_path
+    # ML arguments
+    if ML == True:
+        kws['CUDA'] = CUDA
+        kws['ml_model_path'] = ml_model_path
+        kws['ml_num_attempts'] = ml_num_attempts
```
Comment (alex-rakowski) [inline py4DSTEM/process/diskdetection/diskdetection.py:242]: remove `ml` from the keywords

File: py4DSTEM/process/diskdetection/diskdetection.py [hunk 5]
```diff
@@ -254,6 +273,7 @@ def _get_function_dictionary():
         "dc_GPU_batched" : _find_Bragg_disks_CUDA_batched,
         "dc_dask" : _find_Bragg_disks_dask,
         "dc_ipyparallel" : _find_Bragg_disks_ipp,
+        "dc_ml_ai" : find_Bragg_disks_aiml
```
Comment (alex-rakowski) [inline py4DSTEM/process/diskdetection/diskdetection.py:276]: change to `dc_ml`

File: py4DSTEM/process/diskdetection/diskdetection_aiml.py [hunk 6]
```diff
@@ -10,13 +10,14 @@
...
 import shutil
 import numpy as np
+from py4DSTEM.io.datastructure.py4dstem.qpoints import QPoints
 
 from scipy.ndimage.filters import gaussian_filter
 from time import time
 from numbers import Number
 
-from py4DSTEM.process.utils import get_cross_correlation_fk, get_maxima_2D
-from py4DSTEM.process.diskdetection import universal_threshold
```
Comment (alex-rakowski) [inline py4DSTEM/process/diskdetection/diskdetection_aiml.py:19]: ADD THE IMPORT TO THE IF STATEMENT ~ 436

File: py4DSTEM/process/diskdetection/diskdetection_aiml_cuda.py [hunk 7]
```diff
@@ -26,7 +26,7 @@
 from py4DSTEM.io import PointList, PointListArray
 from py4DSTEM.utils.tqdmnd import tqdmnd
 from py4DSTEM.process.diskdetection.kernels import kernels
-from py4DSTEM.process.diskdetection.diskdetection import universal_threshold
```
Comment (alex-rakowski) [inline py4DSTEM/process/diskdetection/diskdetection_aiml_cuda.py:29]: add to the if global threshold statement

File: py4DSTEM/process/diskdetection/diskdetection_aiml.py [hunk 8]
```diff
@@ -10,13 +10,14 @@
 import json
 import shutil
 import numpy as np
+from py4DSTEM.io.datastructure.py4dstem.qpoints import QPoints
```
Comment (alex-rakowski) [inline py4DSTEM/process/diskdetection/diskdetection_aiml.py:13]: change to Braggvectors

File: py4DSTEM/process/diskdetection/diskdetection_aiml_cuda.py [hunk 9]
```diff
@@ -324,7 +325,7 @@ def _find_Bragg_disks_aiml_single_DP_CUDA(DP, probe,
         peaks = PointList(coordinates=coords)
```
Comment (alex-rakowski) [inline py4DSTEM/process/diskdetection/diskdetection_aiml_cuda.py:325]: change to braggvectors

File: py4DSTEM/process/diskdetection/diskdetection.py [hunk 10]
```diff
@@ -206,7 +216,7 @@ def find_Bragg_disks(
             else:
                 er = f"unrecognized distributed mode {distributed_mode}"
                 raise Exception(er)
-
+    # overwrite if ML selected
```
Comment (bsavitzky) [inline py4DSTEM/process/diskdetection/diskdetection.py:219]: is this comment supposed to be here?

**Review by bsavitzky — APPROVED** [review]
Thanks Alex - this looks great.  That said, I haven't run any tests.  It's passing the flaker, so shouldn't break import, so I'm going to approve and merge this into dev so it's accessible to others for testing - let's make sure to get it tested on a few machines before the next merge into main?

**Pull request merged**